pkg/sql/vars.go

```diff
@@ -398,6 +398,12 @@ var varGen = map[string]sessionVar{
 		Reset: func(_ *sessionDataMutator) error { return nil },
 	},
 
+	// Supported for PG compatibility only.
+	// See https://www.postgresql.org/docs/10/static/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS
+	`max_identifier_length`: {
+		Get: func(evalCtx *extendedEvalContext) string { return "63" },
```

Comment:
We allow identifiers longer than 63 bytes. I think it's more important to return our true limit (is there one? we should set one) than to return the same value as postgres. (what if some silly C program is using this to allocate a fixed buffer size?)

Reply:
> I think it's more important to return our true limit

I don't think we have an identifier length limit and a quick review of our AST scanner didn't find one. @knz are you aware of such a limit?

Regardless, if we don't have a limit (or even if the limit is larger than 63), then isn't this harmless in the same way that reporting a `max_index_keys` variable is harmless. What kinds of problems are you envisioning with reporting a limit beneath what we actually support?

Reply:
> What kinds of problems are you envisioning with reporting a limit beneath what we actually support?

A few possibilities come to mind. The most severe (but fairly implausible) is a C-style program that uses this value to construct a fixed-size buffer, which it then overflows when it sees a longer name. More plausibly, it looks like what activerecord with this value is using it to truncate ruby class names when creating tables. If the limit varied over time, various sorts of mismatches could occur. 

I think we should introduce a limit here, although backwards compatibility might be painful given the way we construct index names. But that's not something we should do in a patch release.

Reply:
I'd be +1 on creating a limit. This would allow us to be a bit more efficient at different levels. I don't think it would hurt too much if we create this limit larger than pg's default.

Reply:
> This would allow us to be a bit more efficient at different levels

Yes this would allow us to conceivably pass identifier names around as arrays on the stack, which would avoid a lot of indirection and forced allocation. It would probably be a nice win overall.

> I think we should introduce a limit here, although backwards compatibility might be painful given the way we construct index names.

The backwards compatibility story is what scares me about this. We'd have to migrate persistent identifiers and it's not clear to me what we would do with persistent identifiers that exceed whatever limit we decide on. What are the best practices for introducing extra data validation when existing data may be in violation?

Reply:
We can migrate clusters that have too-long identifiers by renaming them to a shorter form. (Adding a number in the name if there's a risk of ambiguity). From then on we must find a better algorithm to name auto-generated indexes but that can be done as well.

As to backward-compatibility:

- we have collectively decided earlier this year that it's ok to break backward compatibility with cockroachdb itself a little if that brings us closer to pg. This is an instance of that.

- apps that were developed against postgres would not use long identifiers anyway because they couldn't with pg. This change would affect crdb-specific apps and there are not too many of those yet.

- 63 character is very long. Filesystems have used that as a limit forever and people don't usually complain. I'd be surprised if anyone notices in practice.

We should also implement this as follows: 

- in version 2.1, introduce an error when *creating* new db objects with names that are too long, and a warning in the admin UI if the db already contains names that are too long. And a new algorithm to auto-gen names. But do not restrict the name in storage yet.
- in version 2.2 actually restrict the name in storage.

This way users get a release cycle to adapt.

Thoughts?